### PR TITLE
DRILL-4673: Implement "DROP TABLE IF EXISTS" for drill to prevent FAI…

### PR DESCRIPTION
**Pull request by vdiravka**
…LED status on command return

- implement DROP TABLE IF EXISTS and DROP VIEW IF EXISTS;
- added unit test for DROP TABLE IF EXISTS;
- added unit test for DROP VIEW IF EXISTS;
- added unit test for "IF" hive UDF.

### CODE REVIEW

File: exec/java-exec/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -199,15 +199,17 @@ SqlNode SqlCreateOrReplaceView() :
...
     SqlParserPos pos;
+    boolean viewExistenceCheck = false;
 }
 {
     <DROP> { pos = getPos(); }
     <VIEW>
+    [ <IF> <EXISTS> { viewExistenceCheck = true; } ]
     {
-        return new SqlDropView(pos, CompoundIdentifier());
+        return new SqlDropView(pos, CompoundIdentifier(), viewExistenceCheck);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/codegen/includes/parserImpls.ftl]: Should this be `return new SqlDropViewIfExists(..);`?

File: exec/java-exec/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -245,15 +247,17 @@ SqlNode SqlCreateTable() :
...
     SqlParserPos pos;
+    boolean tableExistenceCheck = false;
 }
 {
     <DROP> { pos = getPos(); }
     <TABLE>
+    [ <IF> <EXISTS> { tableExistenceCheck = true; } ]
     {
-        return new SqlDropTable(pos, CompoundIdentifier());
+        return new SqlDropTable(pos, CompoundIdentifier(), tableExistenceCheck);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/codegen/includes/parserImpls.ftl]: Same here. Should this be `return new SqlDropTableIfExists(..);`?

File: exec/java-exec/src/main/codegen/includes/parserImpls.ftl (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/codegen/includes/parserImpls.ftl]: Please rename the old classes with suffix `..IfExists`.

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java [hunk 3]
```diff
@@ -84,4 +84,10 @@ public void testGetJsonObject() throws Exception {
             "first_name","Bh","last_name","Venkata","position","Store"))
         .go();
   }
+
+  @Test // DRILL-3272
+  public void testIf() throws Exception {
+    printResult(testSqlWithResults("select `if`(1999 > 2000, 'latest', 'old') from hive.kv limit 1"));
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java]: Use `testBuilder(..)` to test for result correctness.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java [hunk 4]
```diff
@@ -69,6 +71,15 @@ public PhysicalPlan getPlan(SqlNode sqlNode) throws ValidationException, RelConv
           .build(logger);
     }
 
+    final Table existingTable = SqlHandlerUtil.getTableFromSchema(drillSchema, tableName);
+
+    if (dropTableNode.checkTableExistance()) {
+      if (existingTable == null || existingTable.getJdbcTableType() != Schema.TableType.TABLE){
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java]: Spacing: `) {`

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java [hunk 5]
```diff
@@ -57,13 +63,20 @@ public SqlOperator getOperator() {
 
   @Override
   public List<SqlNode> getOperandList() {
-    return Collections.singletonList((SqlNode) tableName);
+    List<SqlNode> ops = Lists.newArrayList();
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java]: Use `ImmutableList`

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java [hunk 6]
```diff
@@ -92,4 +105,6 @@ public SqlIdentifier getTableIdentifier() {
     return tableName;
   }
 
+  public boolean checkTableExistance() { return tableExistenceCheck; }
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java]: Spelling: `checkTableExistence` and use multiple lines.

File: exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java [hunk 7]
```diff
@@ -19,14 +19,18 @@
 
 import org.apache.drill.common.exceptions.UserException;
 import org.apache.hadoop.fs.Path;
+import org.junit.Ignore;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java]: Unused?

**Comment by sudheeshkatkam** [comment]
I am unable to review changes to `SqlDropView` and `ViewHandler` classes. They show up as binary files. Please fix?

File: exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java [hunk 8]
```diff
@@ -171,4 +175,45 @@ public void testDropOnImmutableSchema() throws Exception {
 
     Assert.assertTrue("Dropping table on immutable schema failed", dropFailed);
   }
+
+  @Test // DRILL-4673
+  public void testDropTableIfExists() throws Exception {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java]: Please split this into multiple tests.

File: exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestViewSupport.java [hunk 9]
```diff
@@ -591,4 +591,27 @@ public void showProperMsgWhenTryingToDropANonViewTable() throws Exception{
       FileUtils.deleteQuietly(tblPath);
     }
   }
+
+  @Test // DRILL-4673
+  public void testDropViewIfExists() throws Exception {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestViewSupport.java]: Please split into multiple tests.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java [hunk 10]
```diff
@@ -69,6 +71,15 @@ public PhysicalPlan getPlan(SqlNode sqlNode) throws ValidationException, RelConv
           .build(logger);
     }
 
+    final Table existingTable = SqlHandlerUtil.getTableFromSchema(drillSchema, tableName);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java]: Can you confirm that `getTableFromSchema(...)` returns `null` only if the table does not exist?

Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java]: Move this call inside the `if` block.

File: exec/java-exec/src/main/codegen/includes/parserImpls.ftl (hunk 1 above)
Comment (vdiravka) [inline exec/java-exec/src/main/codegen/includes/parserImpls.ftl]: I did accordingly to `SqlCreateView` (without `OrReplace`) but agree with new names. Done.

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java (hunk 3 above)
Comment (vdiravka) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/fn/hive/TestInbuiltHiveUDFs.java]: Done in the new commit.

File: exec/java-exec/src/main/codegen/includes/parserImpls.ftl (hunk 2 above)
Comment (vdiravka) [inline exec/java-exec/src/main/codegen/includes/parserImpls.ftl]: Done in a new commit.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java (hunk 10 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java]: I'm not sure about returning `null ` by `getTableFromSchema(...)` only if the table does not exist. Therefore renamed it to `tableToDrop ` and moved inside `if` block

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java (hunk 4 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableHandler.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java (hunk 5 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java (hunk 6 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDropTable.java]: Done.

**Comment by parthchandra** [comment]
There are two files that are still showing up as binary. Can you update the patch?

File: exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java (hunk 7 above)
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java]: Unused import statement is deleted in the new commit.

File: exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java (hunk 8 above)
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/TestDropTable.java]: Done.

File: exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestViewSupport.java (hunk 9 above)
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestViewSupport.java]: Done.

**Comment by vdiravka** [comment]
@sudheeshkatkam @parthchandra 
I resolved the issue with binary representation of two files. 
The patch is updated. Please check.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java [hunk 11]
```diff
@@ -28,47 +30,57 @@
...
+    String tableName = dropTableIfExistsNode.getName();
     if (drillSchema == null) {
       throw UserException.validationError()
           .message("Invalid table_name [%s]", tableName)
           .build(logger);
     }
 
+    if (dropTableIfExistsNode.checkTableExistence()) {
+      final Table tableToDrop = SqlHandlerUtil.getTableFromSchema(drillSchema, tableName);
+      if (tableToDrop == null || tableToDrop.getJdbcTableType() != Schema.TableType.TABLE) {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java]: Looks like `HbaseSchema#getTable` throws a TableNotFoundException instead of returning `null` when table does not exist.

I am not sure other schema return `null` when table does not exist.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ViewHandler.java [hunk 12]
```diff
@@ -106,36 +106,43 @@ public PhysicalPlan getPlan(SqlNode sqlNode) throws ValidationException, RelConv
...
-        throw UserException.validationError()
-            .message("[%s] is not a VIEW in schema [%s]", viewToDrop, schemaPath)
-            .build(logger);
-      } else if (existingTable == null) {
-        throw UserException.validationError()
-            .message("Unknown view [%s] in schema [%s].", viewToDrop, schemaPath)
-            .build(logger);
+      final Table viewToDrop = SqlHandlerUtil.getTableFromSchema(drillSchema, viewName);
+      if (dropView.checkViewExistence()) {
+        if (viewToDrop == null || viewToDrop.getJdbcTableType() != Schema.TableType.VIEW){
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ViewHandler.java]: Same here. I am not sure a `null` check is sufficient.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java (hunk 11 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java]: Forgot to mention, note that Drill does not support dropping Hbase table. My concern is that DROP TABLE command does not return consistent messages with IF EXISTS clause.

**Comment by julianhyde** [comment]
Why rename DropTable to DropTableIfExists? It's still the DROP TABLE command.

**Comment by sudheeshkatkam** [comment]
I agree; I was confused with the [command definition](https://github.com/apache/drill/pull/541/commits/efbd0df49a6f998900aa9ac0755e1775e1b1bff5#diff-f4825f92440b9aa467d135f04498f406R260) in the parser template (`SqlNode DropTableIfExists():` conflicts with the return statement `return new DropTable(...)`).

**Comment by julianhyde** [comment]
The method in the parser and the class should both be called `DropTable`. `ifExists` should just be a boolean field.

At some point you might want to add an `OR REPLACE` clause to `CREATE VIEW`, and add `IF NOT EXISTS` to `CREATE TABLE`. You would not want to rename those classes, either. Just add a boolean field.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java (hunk 11 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java]: I checked every `schema#getTable` and found that every method can throw exception only if something go wrong (permissionError, dataReadError, IOException etc). 
`getTableFromSchema `can not throw any exception, because in this method `catch (Exception e)` is used.

Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DropTableIfExistsHandler.java]: I checked such query `drop table if exists hbase.TestTableNullStr` and found that the message is proper and same like for `DROP TABLE` statement: `PARSE ERROR: Unable to create or drop tables/views. Schema [hbase] is immutable.`.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ViewHandler.java (hunk 12 above)
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ViewHandler.java]: Answered above.

**Comment by vdiravka** [comment]
Agree with class/method names. I was confused with `SqlCreateOrReplaceView()`. Thanks Julian.
@sudheeshkatkam Could you review please?

**Comment by sudheeshkatkam** [comment]
tl;dr +1, please open a ticket.

Moving conversation. Vitalii said:
> I checked every schema#getTable and found that every method can throw exception only if something go wrong (permissionError, dataReadError, IOException etc). 
getTableFromSchema()can not throw any exception, because in this method catch (Exception e) is used.

Correct, `catch(Exception e)` is used, but the exception is wrapped in a `DrillRuntimeException` and re-thrown.

I think it "happens so" that all plugins are doing the right thing. To explain myself, there are two cases that I am concerned about:

1.  `getTableFromSchema` (and underlying `schema#getTable`) throws an exception when a table does not exist. For example, if the HBase plugin supports dropping tables in the future, an invalid table name will cause the query to [fail with](https://github.com/apache/drill/blob/master/contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/DrillHBaseTable.java#L42) a `TableNotFoundException`, which is exactly the case an exception should not thrown, if `IF EXISTS` exists.  _For now,_ since hbase schema is not mutable, the [check for schema mutability](https://github.com/apache/drill/pull/541/files#diff-50862f72c560b2d1bd9af36e83019d80R66), before deleting the table, fails.

2. `getTableFromSchema` (and underlying `schema#getTable`) does not return `null` for non-existent tables, and `schema#dropTable` will fail. _For now_, only two schema support dropping tables: `WorkspaceSchema` and `KuduTables`, and they return `null` for non-existent tables. The check for schema mutability fails for other schema.

Overall, it seems there is an unwritten rule (`schema#getTable` should return `null` if the table does not exist) that should be documented, and HBase plugin should be fixed. AFAIK other plugins follow this rule.

+1 for the changes; I think we should track these above issues in another ticket.

**Comment by vdiravka** [comment]
The ticket about `schema#getTable` is opened: [DRILL-4799](https://issues.apache.org/jira/browse/DRILL-4799).
Two commits were squashed into one, the branch was rebased to the master version.
